[CORE-13296] Align TLS configuration with policy

BenPope opened the pull request:
## Documentation
### Update

See https://github.com/redpanda-data/redpanda/pull/27835 - the change to the defaults will be deferred until v26.1

### Breaking Changes ###
As of Redpanda v26.1.1 the set of ciphers supported over TLSv1.2 and TLSv1.3 will be reduced and reordered for improved security and performance. The new defaults are all AEAD suites, providing strong authentication, key exchange, forward secrecy, and encryption of at least 128 bits. This applies to all external connections, both incoming and outgoing.

TLS v1.3 should be negotiated for clients running Java 11 or newer (from around Kafka 3.0), or using librdkafka 1.5 (July 2020) or newer, built with OpenSSL 1.0.2 or newer. This should cover the vast majority of use cases.

**Note: Connections between Redpanda brokers are always TLSv1.3, the following configuration options do not affect that.**

To enable support for older clients (such as pre-2018 Java 8), you may set:
```shell
rpk cluster config set \
 tls_v1_2_cipher_suites "ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:AES256-GCM-SHA384" 
```

To revert to the old behaviour (not recommended):
```shell
rpk cluster config set \
 tls_v1_2_cipher_suites "ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:AES256-GCM-SHA384:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:AES128-SHA:AES128-CCM:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:AES256-SHA:AES256-CCM" \
 tls_v1_3_cipher_suites "TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_CCM_SHA256"
```


## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

### Improvements

* Improved default TLS Configuration

BenPope committed:
config: Add TLS cipher configuration

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
config: serde for new tls_config fields

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
config: Override for RPC to internal config

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
net/tls: Expose tls ciphers

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
net/tls: Allow configuration of ciphers

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
config: Add fields to tls_config

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
net/tls: Improve ciphers

The defaults are restricted and ordered by strength.
The use case is for external listeners and clients.

A new strict profile is introduced, intended for internal
communication (currently just internal RPC).

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
dt: Refactor tls_version_test

* Enable TLS on all interfaces
* Allow specifying a cipher

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
dt: Check default ciphers in tls_version_test

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
dt: Check changed ciphers in tls_version_test

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
dt: Check changed min_tls_versions in tls_version_test

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope commented on src/v/config/configuration.cc:
```diff
@@ -3902,6 +3903,18 @@ configuration::configuration()
+  , tls_12_cipher_string(
+      *this,
+      "tls_12_cipher_string",
+      "A colon separated list of IANA cipher strings (compatible with OpenSSL)",
```
Ahhh, actually, these aren't IANA until OpenSSL 3.2

BenPope replied:
Removed the word

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/72155-->
<!--8b6cfa85-49d8-48eb-a9e2-57d4c51ed445-->
<!--bd80b7ae-2ffc-435b-9e58-695cdcecbb5b-->
<!--40c651e2-d42d-49d6-b4c6-b7e32ac5c5c2-->
<!--ab3cb3cd-897f-427f-8e3a-87a7f7e66bf0-->
<!--title-->
## Retry command for [Build#72155](https://buildkite.com/redpanda/redpanda/builds/72155)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/enterprise_features_license_test.py::EnterpriseFeaturesTest.test_enable_features@{"disable_trial":false,"feature":5,"install_license":true}
tests/rptest/tests/enterprise_features_license_test.py::EnterpriseFeaturesTest.test_enable_features@{"disable_trial":true,"feature":5,"install_license":true}
tests/rptest/tests/enterprise_features_license_test.py::EnterpriseFeaturesTest.test_enable_features@{"disable_trial":false,"feature":5,"install_license":false}
tests/rptest/tests/cluster_config_test.py::ClusterConfigTest.test_valid_settings
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--0b6b40ca-515d-4083-9db8-cf983bf94447-->
<!--TEST RESULTS-->
<!--8f9155ef-ca7e-4289-970b-5fa50531e5cf-->
<!--TEST RESULTS-->
<!--b8930976-e53c-418b-b1e0-063fa26b69c4-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72155">build#72155</a></summary>


| test_class             | test_method                          | test_arguments                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-----------------------|:-------------------------------------|:-----------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestOauth      | test_kafka_oauth                     | {"authz_match": "rbac"}                                          | integration | https://buildkite.com/redpanda/redpanda/builds/72155#01993e99-2e61-4915-af8c-7c1c081df586 | FLAKY         | 15/21    | upstream reliability is '91.10512129380054'. current run reliability is '71.42857142857143'. drift is 19.67655 and the allowed drift is set to 50. The test should PASS |
| ClusterConfigTest      | test_valid_settings                  | null                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/72155#01993e9e-23fe-4ecf-9944-01419cbf9546 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DataMigrationsApiTest  | test_creating_and_listing_migrations | null                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/72155#01993e99-2e61-4915-af8c-7c1c081df586 | FLAKY         | 15/21    | upstream reliability is '100.0'. current run reliability is '71.42857142857143'. drift is 28.57143 and the allowed drift is set to 50. The test should PASS             |
| EnterpriseFeaturesTest | test_enable_features                 | {"disable_trial": false, "feature": 5, "install_license": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72155#01993e9e-23f9-471f-b97c-eb0eb743041f | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| EnterpriseFeaturesTest | test_enable_features                 | {"disable_trial": false, "feature": 5, "install_license": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/72155#01993e9e-23fb-4a8c-9579-cf6e5c6e910a | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| EnterpriseFeaturesTest | test_enable_features                 | {"disable_trial": true, "feature": 5, "install_license": true}   | integration | https://buildkite.com/redpanda/redpanda/builds/72155#01993e9e-23fd-4ea2-b461-1618830954bf | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72349">build#72349</a></summary>


| test_class               | test_method                          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:-------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest    | test_creating_and_listing_migrations | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72349#0199531d-b53e-4313-b43c-af07e4dfb6fa | FLAKY         | 20/21    | upstream reliability is '99.29906542056075'. current run reliability is '95.23809523809523'. drift is 4.06097 and the allowed drift is set to 50. The test should PASS |
| DataMigrationsApiTest    | test_migrated_topic_data_integrity   | {"include_groups": true, "params": {"cancellation": {"dir": "in", "stage": "preparing"}, "use_alias": true}, "transfer_leadership": true}      | integration | https://buildkite.com/redpanda/redpanda/builds/72349#0199531d-b53f-4f1e-90eb-6521837c808a | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest | test_node_operations                 | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72349#01995321-98bc-40b1-aea0-06840052f35e | FLAKY         | 19/21    | upstream reliability is '99.58677685950413'. current run reliability is '90.47619047619048'. drift is 9.11059 and the allowed drift is set to 50. The test should PASS |
| SimpleEndToEndTest       | test_consumer_interruption           | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72349#01995321-98b8-43d6-8a58-e99cd13d4870 | FLAKY         | 20/21    | upstream reliability is '98.70801033591732'. current run reliability is '95.23809523809523'. drift is 3.46992 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72402">build#72402</a></summary>


| test_class               | test_method                        | test_arguments                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-------------------------|:-----------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageScrubberTest | test_scrubber                      | {"cloud_storage_type": 1}                                                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/72402#0199554e-57df-4f5e-88ef-ce725c28d49e | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
| DataMigrationsApiTest    | test_migrated_topic_data_integrity | {"include_groups": false, "params": {"cancellation": {"dir": "in", "stage": "prepared"}, "use_alias": false}, "transfer_leadership": true} | integration | https://buildkite.com/redpanda/redpanda/builds/72402#0199554e-57de-4a08-8b31-480a8bd53391 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

IoannisRP commented on src/v/net/tls.cc:
```diff
@@ -41,28 +41,33 @@ ss::future<std::optional<ss::sstring>> find_ca_file() {
+                                               "ECDHE-ECDSA-CHACHA20-POLY1305:"
+                                               "ECDHE-RSA-CHACHA20-POLY1305:"
+                                               "ECDHE-ECDSA-AES128-GCM-SHA256:"
+                                               "ECDHE-RSA-AES128-GCM-SHA256";
```
why not `inline constexpr` variables in the header?
And then you won't have to extern them later on.

BenPope replied:
> why not `inline constexpr` variables in the header? And then you won't have to extern them later on.

It was like that recently, but then was moved to this file, which I'm partially undoing.

The header file is included in a lot of places; it saved me time rebasing and building.

IoannisRP commented on src/v/redpanda/application.cc:
```diff
@@ -2704,9 +2705,29 @@ void application::wire_up_bootstrap_services() {
+                config::node().rpc_server_tls().get_truststore_file(),
+                config::node().rpc_server_tls().get_crl_file(),
+                config::node().rpc_server_tls().get_require_client_auth(),
+                config::node()
+                  .rpc_server_tls()
+                  .get_tls_12_cipher_string()
+                  .value_or(ss::sstring{}),
+                config::node()
+                  .rpc_server_tls()
+                  .get_tls_13_ciphersuites()
+                  .value_or(ss::sstring{net::tlsv1_3_ciphersuites_strict}),
+                config::node().rpc_server_tls().get_min_tls_version().value_or(
+                  config::tls_version::v1_3),
+                config::node()
+                  .rpc_server_tls()
+                  .get_enable_renegotiation()
+                  .value_or(false)};
```
is there an existing codepath that these values can be null? Or is it a safety measure?

Apart from `enable_renegotiation`, the fallback values are different from the cluster config default values. What is the reasoning for it?

BenPope replied:
> is there an existing codepath that these values can be null? Or is it a safety measure?

The values are deseralized from `tls_config`; they can appear in the node config. This is an undocumented "backdoor".

> Apart from `enable_renegotiation`, the fallback values are different from the cluster config default values. What is the reasoning for it?

This RPC is in "internal" connection, and as such has stricter security requirements for TLS.

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/10dc075dc536e9209ba1e12022145c14b1246846..ba3e6615a555e89a0b8bcc88acf669601121358f)
* Rebase
* Dropped the first commit that used the new `net::get_credentials_builder`, to avoid oversize allocs in the test. This can be done as a followup.

pgellert commented on src/v/redpanda/application.cc:
```diff
@@ -2710,9 +2711,29 @@ void application::wire_up_bootstrap_services() {
+                config::node().rpc_server_tls().get_truststore_file(),
+                config::node().rpc_server_tls().get_crl_file(),
+                config::node().rpc_server_tls().get_require_client_auth(),
+                config::node()
+                  .rpc_server_tls()
+                  .get_tls_12_cipher_string()
+                  .value_or(ss::sstring{}),
+                config::node()
+                  .rpc_server_tls()
+                  .get_tls_13_ciphersuites()
+                  .value_or(ss::sstring{net::tlsv1_3_ciphersuites_strict}),
+                config::node().rpc_server_tls().get_min_tls_version().value_or(
+                  config::tls_version::v1_3),
+                config::node()
+                  .rpc_server_tls()
+                  .get_enable_renegotiation()
+                  .value_or(false)};
```
Just double checking: the intended behaviour is that the new cluster configs won't impact the internal RPC.

I think we should document this in either the docs of the cluster config (clarifying what the cluster config impacts and what it doesn't) and/or in our docs page.

BenPope replied:
> Just double checking: the intended behaviour is that the new cluster configs won't impact the internal RPC.

That's correct. The security policy dictates a strict TLS configuration. There should be no need to configure it for old clients, as we only ever support talking Redpanda (give or take a version), which we can enforce to be strict.

> I think we should document this in either the docs of the cluster config (clarifying what the cluster config impacts and what it doesn't) and/or in our docs page.

That's fair.

BenPope replied:
Done

pgellert reviewed (commented):
Just one question, but other than that it looks good to me

BenPope commented on src/v/config/configuration.cc:
```diff
@@ -3941,6 +3942,18 @@ configuration::configuration()
+      "tls_13_ciphersuites",
+      "A colon separated list of ciphersuites (compatible with OpenSSL)",
+      {.needs_restart = needs_restart::yes, .visibility = visibility::user},
+      ss::sstring{net::tlsv1_3_ciphersuites})
```
Suggestion wording:
```
A colon separated list of <cipher strings|suites> (compatible with OpenSSL).
This applies to all external connections, both accepted by and initiated by Redpanda.
```

BenPope replied:
Better suggestion applied.

paulohtb6 commented on src/v/config/configuration.cc:
```diff
@@ -3941,6 +3942,18 @@ configuration::configuration()
+      ss::sstring{net::tlsv1_2_cipher_string})
+  , tls_13_ciphersuites(
+      *this,
+      "tls_13_ciphersuites",
```
btw cipher suites looks to be two separate words per https://en.wikipedia.org/wiki/Cipher_suite.
I would also recommend renaming the property name to something more easily identifiable.

And it looks like both properties do the same thing, but for different versions. I would keep consistency in mind with their names.

Suggested names: 

tls_v12_external_cipher_suites
tls_v13_external_cipher_suites

BenPope replied:
> btw cipher suites looks to be two separate words per https://en.wikipedia.org/wiki/Cipher_suite. I would also recommend renaming the property name to something more easily identifiable.
> 
> And it looks like both properties do the same thing, but for different versions. I would keep consistency in mind with their names.

Well, the naming came from OpenSSL API, but, in hindsight, was a terrible idea.

> Suggested names:
> 
> tls_v12_external_cipher_suites tls_v13_external_cipher_suites

I adorned them a bit.
* tls_v1_2_cipher_suites
* tls_v1_3_cipher_suites

paulohtb6 commented on src/v/config/configuration.cc:
```diff
@@ -3941,6 +3942,18 @@ configuration::configuration()
+      "A colon separated list of cipher strings (compatible with OpenSSL)",
+      {.needs_restart = needs_restart::yes, .visibility = visibility::user},
+      ss::sstring{net::tlsv1_2_cipher_string})
+  , tls_13_ciphersuites(
```
```suggestion
  , tls_13_cipher_suites(
```

paulohtb6 commented on src/v/config/configuration.cc:
```diff
@@ -3941,6 +3942,18 @@ configuration::configuration()
+  , tls_12_cipher_string(
+      *this,
+      "tls_12_cipher_string",
+      "A colon separated list of cipher strings (compatible with OpenSSL)",
```
Not 100% that I got this right. Please review for tech accuracy 

```suggestion
Specifies the TLS 1.2 cipher suites available for external client connections as a colon-separated OpenSSL-compatible list. Configure this property to support legacy clients.
```

paulohtb6 commented on src/v/config/configuration.cc:
```diff
@@ -3941,6 +3942,18 @@ configuration::configuration()
+  , tls_13_ciphersuites(
+      *this,
+      "tls_13_ciphersuites",
+      "A colon separated list of ciphersuites (compatible with OpenSSL)",
```
```suggestion
      "Specifies the TLS 1.3 cipher suites available for external client connections as a colon-separated OpenSSL-compatible list. Most deployments don't need to modify this setting. Configure this property only for specific organizational security policies.",
```

I'm still on the fence why would the user need to set this up. If my current explanation is not right, please check.

BenPope replied:
> I'm still on the fence why would the user need to set this up. If my current explanation is not right, please check.

Yeah, this could probably be `visibility::tunable`.

BenPope commented on src/v/redpanda/application.cc:
```diff
@@ -2710,9 +2711,29 @@ void application::wire_up_bootstrap_services() {
+                config::node()
+                  .rpc_server_tls()
+                  .get_tls_13_ciphersuites()
+                  .value_or(ss::sstring{net::tlsv1_3_ciphersuites_strict}),
```
There's a specific security requirement.

BenPope committed:
config: Add validation of cipher suites

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/f25ee6f5011f7c022fbde23e985a61b59ad86850..d84fed12eee90d5acf97d93169b5938f4d3053b4)
* tls_config
  * Added validation
  * Added some tests for deserialization
* Configuration
  * Renamed the new cluster config
  * Made the description more descriptive
  * Added Validation
* Addressed a review comment from copilot
* Global
  * Renamed members and accessors for consistency

BenPope committed:
pythong fixup refactor

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/d84fed12eee90d5acf97d93169b5938f4d3053b4..f4bb3255755e911f1171d4191560296dd9cbca69)
* Fix type-checking

BenPope commented on src/v/config/tls_config.cc:
```diff
@@ -39,7 +42,24 @@ net::key_store create_key_store(const key_cert_container& container) {
+
+using ssl_ctx_ptr = ssl_handle<SSL_CTX, SSL_CTX_free>;
+
+bool ssl_clean_room(auto func) {
+    auto cleanup = ss::defer([] { ERR_clear_error(); });
+    ssl_ctx_ptr ctx{SSL_CTX_new(TLS_method())};
+    return ctx && func(ctx.get());
+}
```
@michael-redpanda does this seem reasonable and complete?

Do I need to count the stack, and return it to it's original state?

paulohtb6 reviewed (approved):
lgtm from docs
